docs/containerization-design-doc.md

```diff
@@ -0,0 +1,384 @@
+#### Mismatched Queueing from YARN
+Azkaban implements it’s own queueing to maximize the use of bare-metal Executors. This queueing is often
+mismatched from YARN queues causing issues downstream.
+
+#### Deployment Issues
+Update to Job Type code requires a jobtype plugin deployment to happen on all Executors. 
+As part of deployment, the inactive ExecutorServer process may continue for up to 10 days to satisfy Azkaban’s
+SLAs, occupying a lot of memory and CPU. Hence, if a deployment has issues, successive attempts to rectify the
+problem leave additional beefy ExecutorProcesses running, causing memory sap leading to GC pauses and OOMs.
+They also pollute metrics.
+
+## Key Requirements for Containerization
+1. Azkaban Web Server should be able to Launch flows in independent containers, thereby giving a fully
+isolated environment for each flow.
+
+2. Be able to respond quickly in response to Spikes in demand (Flexible Infrastructure).
+
+3. Allow the components that make up Azkaban: Platform/Client binaries for Hadoop, Hive etc., Azkaban itself and
+jobtypes to evolve independently of each other.
+
+   * Give the evolution control for Platform, Azkaban and Jobtypes to their corresponding owners.
+   
+   * Provide users a way to override default binary versions of Azkaban/jobtypes etc. to the version of
+      their choice (Helpful during development process of infrastructure -- Azkaban/jobtypes/platform).
+    
+4. Provide plumbing for a fine-grained Canary system that can allow Azkaban/jobtypes and platform full
+control of ramping up their binaries, independent of each other.
+
+### Future Extensions
+1. Provide the fine-grained Canary system for Multiple components that make up Azkaban to help in their
+independent evolution.
+
+## High Level Design Summary
+![](figures/containerized-high-level-arch.png)
+
+1. Azkaban will follow a **Disposable Container** model. This implies that whenever a flow is to be launched,
+the **dispatch logic** will launch a fresh Pod and the pod is destroyed at the conclusion of the flow.
+
+2. Isolation is achieved per flow (Not at job level). Jobs/subflows that are a part of a flow, will run within the
+confines of the pod launched for orchestrating the flow. Job level isolation was explored and rejected:
+
+   * It is very disruptive given the existing Azkaban architecture. Major portions of executor code will have to
+   be re-written to accomodate job level isolation.
+   
+   * It appears too resource hungry to launch a separate pod per job. Separate container within the same pod is
+   possible. But again, this would have required rewriting major parts of flow - job wiring. This is something that
+   can be reconsidered in future.
+
+3. The pod will be launched with default compute/memory resources, but override parameters will be available
+to request more resources for the flow orchestration.
+
+4. For this design iteration, the web server will stay outside of k8s. This does not preclude the need for
```

Comment:
Is it possible to mention that this design is for any containerized infrastructure? For now, we are adding implementation for Kubernetes.